File: packages/python/src/cci/export.py
"""export() (contracts/operations.md `export()`/`import_history()`; data-model.md Import
validation).

Bounded-stream JSONL: one committed snapshot (a single short read transaction captures the
range up front, then streams — no long-held read transaction), a checksum manifest as the
trailing line. Excludes secrets, memo entries, and private diagnostics by construction — the
memo cache lives in a separate SQLite file/Redis entirely (data-model.md Cache-scoped
entities), never touched here.
"""

from __future__ import annotations

import hashlib
import json
from dataclasses import dataclass

import apsw

from .io_worker import fetchall
from .store import HistoryStore, map_storage_error

EXPORT_FORMAT_VERSION = 1


@dataclass(frozen=True)
class ExportManifest:
    cci_export_version: int
    history_id: str
    message_count: int
    checksum: str


def _message_row_to_record(row: tuple) -> dict:
    (
        message_id, seq, source_id, external_id, idempotency_key, position_in_batch,
        role, original_payload, text_projection, payload_hash, session_metadata, created_at,
    ) = row
    return {
        "record_type": "message",
        "message_id": message_id,
        "seq": seq,
        "source_id": source_id,
        "external_id": external_id,
        "idempotency_key": idempotency_key,
        "position_in_batch": position_in_batch,
        "role": role,
        "original_payload": json.loads(original_payload),
        "text_projection": text_projection,
        "payload_hash": payload_hash,
        "session_metadata": json.loads(session_metadata) if session_metadata else None,
        "created_at": created_at,
    }


async def export(store: HistoryStore, dest_path: str) -> ExportManifest:
    """Streams `store`'s current committed history to `dest_path` as JSONL, one message
    record per line, followed by a trailing manifest line. `write_lock` serializes this read
    transaction against the store's single AsyncConnection like every other transaction here —
    a second concurrent `async with connection:` would otherwise fail
    ("transaction within a transaction")."""
    async with store.write_lock:
        try:
            async with store.connection:
                cursor = await store.connection.execute(
                    "SELECT message_id, seq, source_id, external_id, idempotency_key, "
                    "position_in_batch, role, original_payload, text_projection, "
                    "payload_hash, session_metadata, created_at FROM messages "
                    "WHERE history_id = ? ORDER BY seq",
                    (store.history_id,),
                )
                rows = await fetchall(cursor)
        except apsw.Error as exc:
            raise map_storage_error(exc) from exc

    hasher = hashlib.sha256()
    with open(dest_path, "w", encoding="utf-8") as f:
        for row in rows:
            line = json.dumps(_message_row_to_record(row), sort_keys=True, separators=(",", ":"))
            hasher.update(line.encode("utf-8"))
            f.write(line + "\n")

        manifest = ExportManifest(
            cci_export_version=EXPORT_FORMAT_VERSION,
            history_id=store.history_id,
            message_count=len(rows),
            checksum=hasher.hexdigest(),
        )
        f.write(
            json.dumps(
                {
                    "record_type": "manifest",
                    "cci_export_version": manifest.cci_export_version,
                    "history_id": manifest.history_id,
                    "message_count": manifest.message_count,
                    "checksum": manifest.checksum,
                },
                sort_keys=True,
                separators=(",", ":"),
            )
            + "\n"
        )

    return manifest

File: packages/typescript/src/export.ts
/**
 * export() (contracts/operations.md `export()`/`import_history()`) — mirrors export.py.
 * Bounded-stream JSONL: one committed snapshot, a checksum manifest as the trailing line.
 */

import { createHash } from "node:crypto";
import { writeFileSync } from "node:fs";
import { HistoryStore } from "./store.js";
import { mapStorageError } from "./ioWorker.js";

export const EXPORT_FORMAT_VERSION = 1;

export interface ExportManifest {
  cciExportVersion: number;
  historyId: string;
  messageCount: number;
  checksum: string;
}

function messageRowToRecord(row: Record<string, unknown>): Record<string, unknown> {
  return {
    record_type: "message",
    message_id: row.message_id,
    seq: row.seq,
    source_id: row.source_id,
    external_id: row.external_id ?? null,
    idempotency_key: row.idempotency_key ?? null,
    position_in_batch: row.position_in_batch ?? null,
    role: row.role,
    original_payload: JSON.parse(row.original_payload as string),
    text_projection: row.text_projection ?? null,
    payload_hash: row.payload_hash,
    session_metadata: row.session_metadata ? JSON.parse(row.session_metadata as string) : null,
    created_at: row.created_at,
  };
}

function canonicalLine(obj: Record<string, unknown>): string {
  const sorted: Record<string, unknown> = {};
  for (const key of Object.keys(obj).sort()) sorted[key] = obj[key];
  return JSON.stringify(sorted);
}

export async function exportHistory(store: HistoryStore, destPath: string): Promise<ExportManifest> {
  let rows: Record<string, unknown>[];
  try {
    rows = await store.withLock(() =>
      store.connection.all<Record<string, unknown>>(
        "SELECT message_id, seq, source_id, external_id, idempotency_key, position_in_batch, " +
          "role, original_payload, text_projection, payload_hash, session_metadata, created_at " +
          "FROM messages WHERE history_id = ? ORDER BY seq",
        [store.historyId],
      ),
    );
  } catch (err) {
    throw mapStorageError(err as { message: string; code?: string });
  }

  const hasher = createHash("sha256");
  const lines: string[] = [];
  for (const row of rows) {
    const line = canonicalLine(messageRowToRecord(row));
    hasher.update(line, "utf-8");
    lines.push(line);
  }

  const manifest: ExportManifest = {
    cciExportVersion: EXPORT_FORMAT_VERSION,
    historyId: store.historyId,
    messageCount: rows.length,
    checksum: hasher.digest("hex"),
  };
  lines.push(
    canonicalLine({
      record_type: "manifest",
      cci_export_version: manifest.cciExportVersion,
      history_id: manifest.historyId,
      message_count: manifest.messageCount,
      checksum: manifest.checksum,
    }),
  );

  writeFileSync(destPath, lines.join("\n") + "\n", "utf-8");
  return manifest;
}
